relay/relay.go

```diff
@@ -379,20 +379,17 @@ func (r *Relay) tryRecoverLatestFile(ctx context.Context, parser2 *parser.Parser
 				zap.Stringer("from position", latestPos), zap.Stringer("to position", result.LatestPos), log.WrapStringerField("from GTID set", latestGTID), log.WrapStringerField("to GTID set", result.LatestGTIDs))
 
 			if result.LatestGTIDs != nil {
-				gs := result.LatestGTIDs
-				oldGs1 := gs.Clone()
-				// in MySQL, we expect `PreviousGTIDsEvent` contains ALL previous GTID sets, but in fact it may lack a part of them sometimes,
-				// e.g we expect `00c04543-f584-11e9-a765-0242ac120002:1-100,03fc0263-28c7-11e7-a653-6c0b84d59f30:1-100`,
-				// but may be `00c04543-f584-11e9-a765-0242ac120002:50-100,03fc0263-28c7-11e7-a653-6c0b84d59f30:60-100`.
-				// and when DM requesting MySQL to send binlog events with this EXCLUDED GTID sets, some errors like
-				// `ERROR 1236 (HY000): The slave is connecting using CHANGE MASTER TO MASTER_AUTO_POSITION = 1, but the master has purged binary logs containing GTIDs that the slave requires.`
-				// may occur, so we force to reset the START part of any GTID set.
-				if gs.ResetStart() {
-					r.logger.Warn("force to reset the start part of recovered GTID sets", zap.Stringer("from GTID set", oldGs1), zap.Stringer("to GTID set", gs))
-					oldGs2 := latestGTID.Clone()
-					if latestGTID.ResetStart() {
-						r.logger.Warn("force to reset the start part of latest GTID sets", zap.Stringer("from GTID set", oldGs2), zap.Stringer("to GTID set", latestGTID))
-					}
+				dbConn, err2 := r.db.Conn(ctx)
+				if err2 != nil {
+					return err2
+				}
+				result.LatestGTIDs, err2 = utils.AddGSetWithPurged(ctx, result.LatestGTIDs, dbConn)
+				if err2 != nil {
+					return err2
+				}
+				latestGTID, err2 = utils.AddGSetWithPurged(ctx, latestGTID, dbConn)
```

Comment:
If we don't add purged gtid to `result.LatestGTIDs`, do we need to add purged gtid to `latestGTID` here?